dali/kernels/common/transpose.h

```diff
@@ -0,0 +1,214 @@
+    result.push_back(perm[block.first]);
+    for (int i = block.first + 1; i <= block.second; i++) {
+      removed.push_back(perm[i]);
+    }
+  }
+  for (auto &elem : result) {
+    int original_elem = elem;
+    for (auto removed_elem : removed) {
+      if (original_elem > removed_elem) {
+        elem--;
+      }
+    }
+  }
+  return result;
+}
+
+/**
+ * @brief Based on the blocks returned from PermutationBlocks, collapse the blocks
+ *        in the shape that are permuted together.
+ *
+ * @param shape Shape to be collapsed
+ * @param perm Original, non-collapsed permutation
+ * @param perm_blocks Description of blocks in permutation
+ * @return TensorShape<>
+ */
+TensorShape<> CollapseShape(const TensorShape<> &shape, span<const int> perm,
+                            const std::vector<std::pair<int, int>> &perm_blocks) {
+  TensorShape<> result;
+  result.resize(perm_blocks.size());
+  auto collapsed_perm = CollapsePermutation(perm, perm_blocks);
+  for (size_t i = 0; i < perm_blocks.size(); i++) {
+    result[i] = 1;
+  }
+  for (size_t i = 0; i < perm_blocks.size(); i++) {
+    int source_pos = collapsed_perm[i];
+    for (int j = perm_blocks[i].first; j <= perm_blocks[i].second; j++) {
+      result[source_pos] *= shape[perm[j]];
+    }
+  }
+  assert(volume(result) == volume(shape));
+  return result;
+}
+
+}  // namespace detail
+
+/**
+ * @brief Transpose `src` Tensor to `dst` wrt to permutation `perm`
+ */
+template <typename T>
+void Transpose(const TensorView<StorageCPU, T> &dst, const TensorView<StorageCPU, const T> &src,
+               span<const int> perm) {
+  int N = src.shape.sample_dim();
+  assert(dst.shape.sample_dim() == N);
+  assert(volume(src.shape) == volume(dst.shape));
+  auto dst_strides = GetStrides(dst.shape);
+  auto src_strides = GetStrides(src.shape);
+  VALUE_SWITCH(N, static_dims, (1, 2, 3), (
+    detail::transpose_impl_static<T, static_dims, static_dims>(
+        dst.data, src.data,
+        static_dims, make_span(dst_strides), make_span(src_strides), dst.shape, perm);),
+  (
+    detail::transpose_impl<T, 3>(dst.data, src.data, 0, N,
+        make_span(dst_strides), make_span(src_strides), dst.shape, perm);));
+}
+
+/**
+ * @brief Transpose `src` Tensor to `dst` wrt to permutation `perm`
+ *
+ * Internally collapse the groups of consecutive dimensions for the transpositon
+ *
+ * For example "HWC", perm = {2, 0, 1}; the "HW" would be collapsed to one dimension "X",
+ * and effectively we will do XC -> CX transposition.
+ */
+template <typename T>
+void TransposeGrouped(const TensorView<StorageCPU, T> &dst,
+                      const TensorView<StorageCPU, const T> &src, span<const int> perm) {
+  int N = src.shape.sample_dim();
+  auto src_shape = src.shape;
+  auto perm_blocks = detail::PermutationBlocks(perm);
+  auto collapsed_src_shape = detail::CollapseShape(src_shape, perm, perm_blocks);
+  auto collapsed_perm = detail::CollapsePermutation(perm, perm_blocks);
```

Comment:
This operation is done twice - here and inside CollapseShape. Maybe collapsing and permuting the shape should be separated.

Reply:
Now the CollapseShape does only shape-collapsing without using the CollapsePermutation.